src/groupeddataframe/splitapplycombine.jl

```diff
@@ -741,74 +775,45 @@ function _combine(f::AbstractVector{<:Pair}, gd::GroupedDataFrame)
+               agg = check_aggregate(last(p))
+               outcol = agg(incol, gd)
+               res[i] = idx_agg, outcol
+           else
+               fun = do_f(last(p))
+               if p isa Pair{<:ColumnIndex}
+                   incols = gd.parent[!, first(p)]
+               else
+                   df = gd.parent[!, first(p)]
+                   incols = Tuple(eachcol(df))
+               end
+               firstres = do_call(fun, gd.idx, gd.starts, gd.ends, gd, incols, 1)
+               idx, outcols, _ = _combine_with_first(wrap(firstres), fun, gd, incols)
+               res[i] = idx, outcols[1]
+           end
+      end
     # TODO: avoid recomputing idx for each pair when !all(isagg, f)
     idx = res[1][1]
     outcols = map(x -> x[2], res)
     if !all(x -> length(x) == length(outcols[1]), outcols)
         throw(ArgumentError("all functions must return values of the same length"))
     end
-    nams = [f[i] isa Pair{<:ColumnIndex} ?
-                Symbol(names(gd.parent)[index(gd.parent)[first(f[i])]],
-                       '_', funname(last(f[i]))) :
-                Symbol('x', i)
-            for i in 1:length(f)]
-    valscat = DataFrame(collect(AbstractVector, outcols), nams, makeunique=true)
+    valscat = DataFrame(collect(AbstractVector, outcols))
     return idx, valscat
 end
 
-function _combine(f::Any, gd::GroupedDataFrame)
-    if f isa Pair{<:ColumnIndex}
-        incols = gd.parent[!, first(f)]
-        fun = last(f)
-    elseif f isa Pair
-        df = gd.parent[! , collect(first(f))]
-        incols = NamedTuple{Tuple(names(df))}(eachcol(df))
-        fun = last(f)
-    else
-        incols = nothing
-        fun = f
-    end
-    agg = check_aggregate(fun)
-    if agg isa AbstractAggregate && f isa Pair && first(f) isa ColumnIndex
-        idx = Vector{Int}(undef, length(gd))
-        fillfirst!(nothing, idx, 1:length(gd.groups), gd)
-        outcols = (agg(incols, gd),)
-        # nms is set below
-    else
-        firstres = do_call(fun, gd.idx, gd.starts, gd.ends, gd, incols, 1)
-        idx, outcols, nms = _combine_with_first(wrap(firstres), fun, gd, incols)
-    end
-    if f isa Pair && first(f) isa ColumnIndex &&
-        (agg isa AbstractAggregate ||
-         !isa(firstres, Union{AbstractDataFrame, NamedTuple, DataFrameRow, AbstractMatrix}))
-         nms = [Symbol(names(gd.parent)[index(gd.parent)[first(f)]], '_', funname(fun))]
-    end
+function _combine(fun::Function, gd::GroupedDataFrame)
+    idx, outcols, nms = _combine_with_first(wrap(fun(gd[1])), fun, gd, nothing)
     valscat = DataFrame(collect(AbstractVector, outcols), nms)
     return idx, valscat
 end
 
 function _combine_with_first(first::Union{NamedTuple, DataFrameRow, AbstractDataFrame},
                              f::Any, gd::GroupedDataFrame,
-                             incols::Union{Nothing, AbstractVector, NamedTuple})
+                             incols::Union{Nothing, AbstractVector, Tuple})
```

Comment:
Also adjust comment below: "# NamedTuple giving a single row".

Reply:
The comment refers to `first` not to `incols` (actually you have written this comment). (I leave it unresolved as maybe I did not understand something)